sgkit/tests/test_pc_relate.py

```diff
@@ -0,0 +1,127 @@
+    with pytest.raises(ValueError, match=r"MAF must be between \(0.0, 1.0\)"):
+        pc_relate(g, maf=-1)
+    with pytest.raises(ValueError, match=r"MAF must be between \(0.0, 1.0\)"):
+        pc_relate(g, maf=1.0)
+    with pytest.raises(ValueError, match=r"MAF must be between \(0.0, 1.0\)"):
+        pc_relate(g, maf=0.0)
+
+
+@given(arrays(np.int8, (3, 5)))
+@settings(max_examples=10)
+def test_gramian_is_symmetric(a: np.ndarray) -> None:
+    b = gramian(a)
+    assert np.allclose(b, b.T)
+
+
+def test_collapse_ploidy() -> None:
+    g = simulate_genotype_call_dataset(1000, 10, missing_pct=0.1)
+    assert g.call_genotype.shape == (1000, 10, 2)
+    assert g.call_genotype_mask.shape == (1000, 10, 2)
+
+    # Sprinkle some tests data, this is a bit verbose, but in tests verbosity is not bad:
+    g.call_genotype.loc[dict(variants=1, samples=1, ploidy=0)] = 1
+    g.call_genotype.loc[dict(variants=1, samples=1, ploidy=1)] = 1
+    g.call_genotype_mask.loc[dict(variants=1, samples=1, ploidy=0)] = 0
+    g.call_genotype_mask.loc[dict(variants=1, samples=1, ploidy=1)] = 0
+
+    g.call_genotype.loc[dict(variants=2, samples=2, ploidy=0)] = 0
+    g.call_genotype.loc[dict(variants=2, samples=2, ploidy=1)] = 1
+    g.call_genotype_mask.loc[dict(variants=2, samples=2, ploidy=0)] = 0
+    g.call_genotype_mask.loc[dict(variants=2, samples=2, ploidy=1)] = 0
+
+    g.call_genotype.loc[dict(variants=3, samples=3, ploidy=0)] = -1
+    g.call_genotype.loc[dict(variants=3, samples=3, ploidy=1)] = 1
+    g.call_genotype_mask.loc[dict(variants=3, samples=3, ploidy=0)] = 1
+    g.call_genotype_mask.loc[dict(variants=3, samples=3, ploidy=1)] = 0
+
+    call_g, call_g_mask = _collapse_ploidy(g)
+    assert call_g.shape == (1000, 10)
+    assert call_g_mask.shape == (1000, 10)
+    assert call_g.isel(variants=1, samples=1) == 2
+    assert call_g.isel(variants=2, samples=2) == 1
+    assert call_g.isel(variants=3, samples=3) == -1
+    assert call_g_mask.isel(variants=1, samples=1) == 0
+    assert call_g_mask.isel(variants=3, samples=3) == 1
+
+
+def test_impute_genotype_call_with_variant_mean() -> None:
+    g = simulate_genotype_call_dataset(1000, 10, missing_pct=0.1)
+    call_g, call_g_mask = _collapse_ploidy(g)
+    # Sprinkle some tests data
+    call_g.loc[dict(variants=2)] = 1
+    call_g.loc[dict(variants=2, samples=1)] = 2
+    call_g_mask.loc[dict(variants=2)] = False
+    call_g_mask.loc[dict(variants=2, samples=[0, 9])] = True
+    imputed_call_g = _impute_genotype_call_with_variant_mean(call_g, call_g_mask)
+    assert imputed_call_g.isel(variants=2, samples=1) == 2
+    assert (imputed_call_g.isel(variants=2, samples=slice(2, 9)) == 1).all()
+    assert (imputed_call_g.isel(variants=2, samples=[0, 9]) == (7 + 2) / 8).all()
+
+
+def test_pc_relate__values_within_range() -> None:
+    n_samples = 100
+    g = simulate_genotype_call_dataset(1000, n_samples)
+    call_g, _ = _collapse_ploidy(g)
+    pcs = PCA(call_g, ncomp=2).loadings
```

Comment:
I took a look at the satsmodels code/docs and this would give the eigenvectors from a decomposition on input data where the samples are features rather than observations.  It happens to have the same dimensions but is not the same as the projection of the samples into the PC space implied by a decomposition on the transposed inputs (i.e. samples x variants rather than variants x samples like `call_g`).  It's confusing because sometimes people refer to those projections as "loadings" as well, but "scores" is more common.  And the `pcs` should be scores rather than loadings -- I haven't checked GENESIS but I took a look at Hail to be certain.  It also doesn't really make sense to use scaled eigenvectors from the transposed problem for anything unless those eigenvectors come from left-singular vectors instead of right-singular vectors (they'll be right singular when coming from `PCA`).

I would suggest using either one of these to generate scores instead:

```python
# `x` is should have shape (samples, variants) for us

def pc_scores_numpy(x, k):
    x = x - x.mean(axis=0, keepdims=True)
    u, s, v = np.linalg.svd(x, full_matrices=False)
    r = u * s[np.newaxis]
    return r[:, :k]

def pc_scores_sklearn(x, k):
    return PCA(n_components=k, svd_solver='full').fit_transform(x)

# Then in the test:
# pcs = pc_scores(call_g.T.values, 2)
# ds["sample_pcs"] = (("components", "samples"), pcs.T)
```

This was all fresh since I was doing a lot of work with SVD this week, but I wanted to double check the above as well as the usual "loadings" definitions so I verified in this gist: https://gist.github.com/eric-czech/52a04007f4cfb8f05b312079ca770add

Reply:
Oh also I found that the statsmodels.PCA was super slow even with small data, so those definitely speed things up a lot.

Reply:
Oh dang, I remember I meant to validate that, thanks for the catch! Also I did not want to add sklearn as another dependency, but I guess since it's dev/test deps, I guess that was a premature optimisation on my part. Fixed, used sklearn.